### PR TITLE
Nested model bugs

### LINKED ISSUE #1127: [BUG] Nested Schemas throws error `Expected nestedFoo to be of type object, instead found type object`
### Summary:
When defining a schema that references another model, there is a typescript error, and the nested document is resolved before calling `populate()` (see code below)

~It is a bit unclear to me based on the documentation [found here](https://dynamoosejs.com/guide/Schema#model-types), whether I am supposed to use the Dog Schema definition or the Dog Model definition when defining the owner schema. The docs lead me to believe I need to specify the Model definition, but there is a typescript error when doing so, and if I override the typescript error, it doesn't properly set the nested property. Using the Dog Schema when defining the Owner Schema, there are no typescript errors, but there is a runtime error as described in the title and below.~

Upon trying to add the Dog model to the Owner Schema, I get a type error. Based on [@fishcharlie's comment below](https://github.com/dynamoose/dynamoose/issues/1127#issuecomment-782902525) it seems like this is how one would have a document reference another (ie, two tables, one with a link to another ).

I did a sample using a `Dynamoose.THIS` self reference and it worked like what I imagine a schema referencing another model would work (for example, calling populate to populate the nested document data).

### Code sample:
I have tested with the following simple example, similar to the one provided in the documentation, but with different objects

#### Schema
```typescript
const DogSchema = new dynamoose.Schema({
	id: String,
	name: String,
});

const OwnerSchema = new dynamoose.Schema({
	id: String,
	name: String,
	dog: Dog, // defined below as a model.
});
```

#### Model
```typescript
const Dog = dynamoose.model('Dog', DogSchema);

const Owner = dynamoose.model('Owner', OwnerSchema);
```

#### General
```typescript
        const fido = new Dog({
		id: 'fidoId',
		name: 'Fido',
	});
	const greg = new Owner({
		id: 'gregId',
		name: 'Greg',
		dog: fido,
	});

	await fido.save();
	const gregSaved = await greg.save();
	console.log({ gregSaved });
	const fetchedGreg = await Owner.get('gregId');
	console.log({ fetchedGreg });
	const populatedGreg = await fetchedGreg.populate();
	console.log({ populatedGreg });
```

### Current output and behavior (including stack trace):

Code fails to compile, Typescript error as follows:

```
Type 'ModelType<AnyDocument>' is not assignable to type 'string | symbol | Schema | StringConstructor | BooleanConstructor | NumberConstructor | typeof Buffer | ... 6 more ... | AttributeDefinition[]'.
  Type 'AnyDocument & Model<AnyDocument> & ModelDocumentConstructor<AnyDocument>' is missing the following properties from type 'StringConstructor': fromCharCode, fromCodePoint, rawts(2322)
Schema.d.ts(86, 5): The expected type comes from this index signature.
```

When bypassing the typescript error (typing `as any`) and running the code anyways, the `fetchedGreg` object is returning the following, with the dog Document populated

```
{
  fetchedGreg: Document {
    name: 'Greg',
    dog: Document {
      name: 'Fido',
      id: 'fidoId'
    },
    id: 'gregId'
  }
}
```

### Expected output and behavior:

**1. There is no typescript error when specifying a Model in another Schema.**

2. With the error solved, the `Owner.get('id')` operation returns the following:
```
{
  fetchedGreg: Document {
    name: 'Greg',
    dog: 'fidoId',   
    id: 'gregId'
  }
}
```

When populate() is called, the following is returned:

```
{
  populatedGreg: Document {
    name: 'Greg',
    dog: Document {
      name: 'Fido',
      id: 'de8aa0b3-8c53-4e53-8e7f-2ebf67ceb0fa'
    },
    id: 'fab2bac3-8eaa-435c-8406-b2fd25cfdd32'
  }
}
```

I expect that to be the output because that is how dynamoose.THIS works (nested document is an ID until you use `populate()`) (Based on the exact example in the [documentation](https://dynamoosejs.com/guide/Schema#model-types))

### Environment:

Operating System: Windows 10 WSL Ubuntu 20.04 LTS
Operating System Version: WSL 2 Ubuntu 20.04 LTS
Node.js version (`node -v`): v12.19.1
NPM version: (`npm -v`): 7.5.4
Dynamoose version: 2.7.0

### Other information (if applicable):

If I have misunderstood the documentation I can submit a PR to try and add some clarification for future users. If is a typing bug I would be happy to try and fix it as well. If it is a general bug, I might not have the time to figure out and track down the issue in the code. 

### Other:
- [x] I have read through the Dynamoose documentation before posting this issue
- [x] I have searched through the GitHub issues (including closed issues) and pull requests to ensure this issue has not already been raised before
- [x] I have searched the internet and Stack Overflow to ensure this issue hasn't been raised or answered before
- [x] I have tested the code provided and am confident it doesn't work as intended
- [x] I have filled out all fields above
- [x] I am running the latest version of Dynamoose

### PR DESCRIPTION
**Pull request by fishcharlie**
### Summary:

This PR fixes a few TypeScript bugs related to nested models, and an issue where nested models would auto-populate sometimes.

### GitHub linked issue:

Closes #1127 

### Type (select 1):
- [x] Bug fix
- [ ] Feature implementation
- [ ] Documentation improvement
- [ ] Testing improvement

- [ ] Test added to report bug (GitHub issue #--- @---)
- [ ] Something not listed here

### Is this a breaking change? (select 1):
- [ ] 🚨 YES 🚨
- [x] No
- [ ] I'm not sure

### Is this ready to be merged into Dynamoose? (select 1):
- [x] Yes
- [ ] No

### Are all the tests currently passing on this PR? (select 1):
- [x] Yes
- [ ] No

### Other:
- [x] I have read through and followed the Contributing Guidelines
- [x] I have searched through the GitHub pull requests to ensure this PR has not already been submitted
- [x] I have updated the Dynamoose documentation (if required) given the changes I made
- [x] I have added/updated the Dynamoose test cases (if required) given the changes I made
- [x] I have ensured the following commands are successful from the root of the project directory
  - [x] `npm test`
  - [x] `npm run lint`
- [x] I agree that all changes made in this pull request may be distributed and are made available in accordance with the [Dynamoose license](https://github.com/dynamoose/dynamoose/blob/master/LICENSE)
- [x] All of my commits and commit messages are detailed, explain what changes were made, and are easy to follow and understand
- [x] I have filled out all fields above

### CODE REVIEW

**Comment by fishcharlie** [comment]
@BraidenCutforth Please review and test this PR and make sure it fixes your issue you had relating to #1127.

**Comment by BraidenCutforth** [comment]
@fishcharlie I have tested this against the original bug I filed for a single nested document. **This works as expected.**

I am not sure if I found a new bug relating to this ticket, or if the bug is considered part of this ticket. The issue is that arrays don't work, and do not throw an error.  My [comment on the original issue](https://github.com/dynamoose/dynamoose/issues/1127#issuecomment-788486135) shows the exact scenario I tested.

If this is intended to be supported I can either create a new bug or add it to my existing issue and test against this PR.

File: test/unit/Model.js [hunk 1]
```diff
@@ -1476,6 +1476,161 @@ describe("Model", () => {
...
+							"name": "Charlie",
+							"parent": 2
+						});
+						expect(getItemTimesCalled).to.eql(1);
+					});
+
+					// TODO: reeneable the tests below for populate with sets
+					/*
+					it("Should not populate document automatically when using set", async () => {
+						let getItemTimesCalled = 0;
```
Comment (BraidenCutforth) [inline test/unit/Model.js:1558]: Based on this, I am assuming the sets/arrays use case is not going to be part of this ticket?

**Review by BraidenCutforth — COMMENTED** [review]
Code looks good, just want to connect about my other comments before I approve.

File: test/types/Schema.ts [hunk 2]
```diff
@@ -62,6 +62,11 @@ const shouldSucceedWithConstantType = new dynamoose.Schema({
 	}
 });
 
+const shouldSuccessWithModelAsType = new dynamoose.Schema({
```
Comment (fishcharlie) [inline test/types/Schema.ts:65]:
```suggestion
const shouldSucceedWithModelAsType = new dynamoose.Schema({
```

File: test/unit/Model.js (hunk 1 above)
Comment (fishcharlie) [reply test/unit/Model.js:1558]: @BraidenCutforth There was an issue here. I forget exactly what it was. Let me circle back and see what I can do.

**Comment by fishcharlie** [comment]
@BraidenCutforth Just pushed some changes that should address this. Wanna take another look and see how it works for you?

**Comment by BraidenCutforth** [comment]
Okay, I have tested this and I was able to get it to work in what seems like a very specific way, with help of your tests. 

Here is what I have come up with:

```typescript
// My setup
const DogSchema = new dynamoose.Schema({
	id: String,
	name: String,
});
const Dog = dynamoose.model('Dog', DogSchema);

const OwnerSchema = new dynamoose.Schema({
    // See below for different variants and their results
})

const Owner = dynamoose.model('Owner', OwnerSchema);
```

The following works for defining a nested model in schema, but **DOES NOT WORK when calling populate**

```typescript
const OwnerSchema = new dynamoose.Schema({
	id: String,
	name: String,
	dog: {
		type: Set,
		schema: [{ type: Dog }], // I believe this implies an array of types, which may explain why populate doesn't work
	},
});
```

The following scenarios **do not work at all**, but from my understanding from the docs, they should:

````typescript
// TypeError: Cannot read property 'toLowerCase' of undefined
// When trying to save a Owner with nested dogs
const OwnerSchema = new dynamoose.Schema({
	id: String,
	name: String,
	dog: {
		type: Set,
		schema: Dog,
	},
});
```

```typescript
// TypeError: Cannot read property 'toLowerCase' of undefined
// When trying to save a Owner with nested dogs
const OwnerSchema = new dynamoose.Schema({
	id: String,
	name: String,
	dog: {
		type: Set,
		schema: { type: Dog },
	},
});
```

Additionally, the array type does not work in any instance. Granted in our scenario a set makes more sense so that is what I will be using, but am under the impression Array type is supposed to work.

@fishcharlie I am unsure of what the scope of this ticket is, if it is just to fix my problem, or fix all problems in relation to nested models.

**Comment by fishcharlie** [comment]
@BraidenCutforth 

> The following scenarios do not work at all, but from my understanding from the docs, they should

`schema: Dog` and `schema: { type: Dog }` should **not** work. So that is expected behavior. Where in the documentation does it indicate that it should work? I wanna get that updated to be more clear.

> `schema: [{ type: Dog }]`

Should work. And implies a set of Dog items/documents. I will investigate why this isn't working when I get a chance.

> I am unsure of what the scope of this ticket is, if it is just to fix my problem, or fix all problems in relation to nested models.

Honestly, I'm not really sure either. Right now, it's to fix all known problems in relation to nested models. At some point I might cut it short and merge it in. But I really appreciate the bug reports. It's really helpful to the project. So I'm taking advantage of your amazing reports to help fix these bugs and make the project better.

**Comment by BraidenCutforth** [comment]
Sorry, upon further reading of the docs, it seems none of those should be supported. I thought `schema: Dog` would work becuase you can specify `dog: Dog` without the array like syntax. Most of my reading in the docs for this specifically has been around [this section](https://dynamoosejs.com/guide/Schema/#model-types). 

I just tested `{ type: Set, schema: [Dog]}` and that worked (barring the issue around the populate command).

**The only issue I currently see then is the populate command not working to populate the nested documents** 

Now this is definitely outside the original issue that was filed, but I feel like it would be nice to have. I can file another bug to close up this ticket or it can tag along, or you can even tell me its not meant to work right now. In any case, let me know what you think about that. Once you get back to me I will review based on what you think should be accomplished in this ticket.

**Comment by fishcharlie** [comment]
@BraidenCutforth Try this out and see how it works for you.

**Comment by BraidenCutforth** [comment]
@fishcharlie this works as expected now! Thank you!

**Review by BraidenCutforth — APPROVED** [review]

**Review by BraidenCutforth — APPROVED** [review]

File: lib/Populate.ts [hunk 3]
```diff
@@ -29,7 +29,7 @@ export function PopulateDocument (this: Document, settings?: PopulateSettings |
 	const {model} = this;
 	const localSettings = settings;
 	const promise = model.schemaForObject(this).then((schema) => {
-		const modelAttributes: any[] = utils.array_flatten(schema.attributes().map((prop) => ({prop, "details": schema.getAttributeTypeDetails(prop)}))).filter((obj) => Array.isArray(obj.details) ? obj.details.some((detail) => detail.name === "Model") : obj.details.name === "Model").map((obj) => obj.prop);
+		const modelAttributes: any[] = utils.array_flatten(schema.attributes().map((prop) => ({prop, "details": schema.getAttributeTypeDetails(prop)}))).filter((obj) => Array.isArray(obj.details) ? obj.details.some((detail) => detail.name === "Model"/* || detail.name === "Model Set"*/) : obj.details.name === "Model" || obj.details.name === "Model Set").map((obj) => obj.prop);
```
Comment (BraidenCutforth) [inline lib/Populate.ts:32]: Remove the commented out part?

**Review by BraidenCutforth — APPROVED** [review]
Sorry, meant to put this with the original review.

File: lib/Populate.ts (hunk 3 above)
Comment (fishcharlie) [reply lib/Populate.ts:32]: @BraidenCutforth I should just add a TODO here. Technically I think this needs to be done. But I can't think of how to build a test around it right now. It's such a weird & unique edge case, that I think it's ok for now.

Comment (fishcharlie) [reply lib/Populate.ts:32]:
```suggestion
		// TODO: uncomment out `/* || detail.name === "Model Set"*/` part and add relevant tests
		const modelAttributes: any[] = utils.array_flatten(schema.attributes().map((prop) => ({prop, "details": schema.getAttributeTypeDetails(prop)}))).filter((obj) => Array.isArray(obj.details) ? obj.details.some((detail) => detail.name === "Model"/* || detail.name === "Model Set"*/) : obj.details.name === "Model" || obj.details.name === "Model Set").map((obj) => obj.prop);
```
